Update GLSP to next (pre-release) version and remove custom code

martin-fleck-at opened the pull request:
- Remove custom IvyUIExtension and replace with GLSPAbstractUIExtension
- Remove custom CSS helper functions (but leave test case for now)

- Use optional grid module from GLSP
-- Replaces setting of custom grid snapper
-- Provides it's own ShowGridAction
-- Uses GridManager to manage visibility
-- Renders grid directly in the graph view
-- CHANGE: Also zoom dot in pattern (1px to zoom size)

- Remove customizations of change bounds tool
-- Add feedback centralized in custom ChangeBoundsManager
-- Add action negative area feedback action (ShowNegativeAreaAction)

- Remove lane change bounds tool and re-use change bounds tool
-- Set resize location to Top and Bottom for lanes
-- Provide custom view for resize handles to render lane handles

Known Issues:
- Insert position of ghost element: server seems to re-center?

ivy-cst commented:
@martin-fleck-at Looks like it is not mergable right now.

martin-fleck-at commented:
@ivy-cst Sorry for that! Rebased the change on top of the current master.

ivy-lli reviewed (commented):
Not finished yet (didn't had much time for review yet).

Found problems:
- ghost element -> as you mentioned
- SVG export
  - seems no longer to work, but also with a normal 11.3 release. So I assume that it broke with the update to GLSP 2 last time 😬 
  - but also something really strange happens when I try to export it when I fit to screen on a specific element (seems to come from this update):
  
![Screen Recording 2024-06-04 at 16 11 09](https://github.com/axonivy/process-editor-client/assets/42733123/6f8adb05-e432-4f37-b2c6-f6f3b9fce501)

ivy-lli commented on packages/editor/src/diagram/activities/activity-views.tsx:
```diff
@@ -14,28 +14,30 @@ const JSX = { createElement: svg };
     return (
       <g>
         <rect
+          attrs={{ [ATTR_BBOX_ELEMENT]: true }}
```
what is this used for?

martin-fleck-at replied:
If that attribute is set, the bounds of this rect-element are being used to calculate the bounds of the element, e.g., ActivityNode. When we need to determine the sizes of the elements, they are rendered on the hidden graph of sprotty without any zooming etc. applied. During that [hidden rendering](https://github.com/eclipse-sprotty/sprotty/blob/8eb479fbd7b1836d32ad82d7a8188769a474598c/packages/sprotty/src/features/bounds/hidden-bounds-updater.ts#L158), we query the browser for the bounding box of the rendered element ([getBBox](https://developer.mozilla.org/en-US/docs/Web/API/SVGGraphicsElement/getBBox)) and apply that as size to the element. However, you can run into problems when the rendering of the element and the bounds of an element get in "conflict" with each other, e.g., when rendering children outside the bounds as decoration or - as I have learned - when rendering the diamond node which for some reason produces a different bounding box than expected. For this reason, the hidden bounds updater allows us to specify that attribute and if an element with that attribute is present, we are taking the bounds of that SVG element as the bounds for the GLSP element.

ivy-lli commented on packages/editor/src/diagram/activities/activity-views.tsx:
```diff
@@ -14,28 +14,30 @@ const JSX = { createElement: svg };
 
-  render(node: ActivityNode, context: RenderingContext): VNode {
+  render(node: ActivityNode, context: RenderingContext): VNode | undefined {
+    if (!this.isVisible(node, context)) {
+      return undefined;
+    }
```
if this is a performance improvement, nice :)

martin-fleck-at commented on packages/editor/src/di.config.ts:
```diff
@@ -94,32 +87,24 @@ export default function createContainer(widgetId: string, ...containerConfigurat
-    minimumMoveDelta: { x: IvyGridSnapper.GRID.x * 2, y: IvyGridSnapper.GRID.y * 2 },
-    alignmentElementFilter: element =>
-      !(element instanceof LaneNode) &&
-      !(element.id === feedbackEdgeId(element.root)) &&
-      !(element.id === feedbackEdgeEndId(element.root)) &&
```
Yes, those two are now part of the default since they are only part of the feedback: https://github.com/eclipse-glsp/glsp-client/blob/65b68e37b8d18852c6a8a4a81f5db5f716ef926b/packages/client/src/features/helper-lines/helper-line-feedback.ts#L75

ivy-lli commented on packages/editor/src/diagram/model.ts:
```diff
@@ -56,10 +57,14 @@ export class LaneNode extends RectangularNode implements WithEditableLabel, Args
   ];
 
   args: Args;
+  get resizeLocations(): ResizeHandleLocation[] {
+    // ideally the server already sets these
+    return [ResizeHandleLocation.Top, ResizeHandleLocation.Bottom];
+  }
```
cool 👍

ivy-lli reviewed (commented):
not finished yet...

martin-fleck-at commented:
@ivy-lli I'll have a look at the SVG export to see what is going on. We do have a customization here so maybe that is causing problems as well. In general, the `isVisible` optimization is not applied on the hidden rendering (which we need to determine sizes) and the SVG export works on that. So I assume something there might be causing those issues.

martin-fleck-at commented:
@ivy-lli I pushed another commit that should fix the SVG exporter. I also added the standalone export module that triggers the export with `Ctrl+Shift+E`.

ivy-lli commented on packages/editor/src/tools/ivy-change-bounds-manager.ts:
```diff
@@ -0,0 +1,46 @@
+
+  protected hideQuickUIFeedback(feedback: FeedbackEmitter): void {
+    feedback.add(
+      SetUIExtensionVisibilityAction.create({ extensionId: QuickActionUI.ID, visible: false }),
+      SetUIExtensionVisibilityAction.create({
+        extensionId: QuickActionUI.ID,
+        visible: true,
+        contextElementsId: [...this.selectionService.getSelectedElementIDs()]
+      })
```
is the second to show the quickaction UI again after the move/resize

martin-fleck-at replied:
Exactly! We introduced this feedback emitter helper object and now when you add a feedback action you can already specify how that action should be cleaned up as soon as the feedback gets disposed. So in this case, during move or resize we hide but as soon as we no longer move/resize and dispose that feedback we show the quickaction UI again. Same for the negative area: We show it during move/resize but hide it again after we are done.

ivy-lli reviewed (approved):
Looks great! Just some questions:

ivy-lli commented:
Or maybe one last thing which I just checked: 
The default size of our activities is currently 60x112, before it was possible to resize them also to this size with the grid snapper. Now it will snap on 62x112 or 54x112...

I'm unsure what we have discussed on this, and I'm also not sure if this is a problem... Maybe we should rather implement a method to reset the size to the default, as also the helper lines can prevent you from resizing the element to "uncommon sizes".

Any thoughts @ivy-cst?

martin-fleck-at commented:
@ivy-lli Thanks! I need to push another update to GLSP but then I'll update this branch with the latest version and we should be good to go.

As for the sizing to uncommon sizes: You still have full control over resizing when you hold `Shift` as this disables the snapping. You can also resize to a non-snapped size if that size is defined as the minimum size of the element, i.e., if you want to get smaller but we hit the minimum, we accept the minimum. Alternatively we could adapt the snapper to snap differently for the resize as we do get the resize handle in but the handle has no bounds so we simply snap it on the grid. I'm quite confident that we can find a good solution if we know how the behavior should be. Please note that when you hit `Ctrl` you now also have the option to do a symmetric resize but we could also activate that as the default.

martin-fleck-at commented:
@ivy-cst @ivy-lli I resolved the merge conflict and updated again to the latest nightly due to the Sprotty upgrade. Do we already have a decision regarding the resizing?

ivy-lli commented:
@martin-fleck-at The ghost element is now fixed from the server side. 
The 2 tests are now run slightly further but still have some problems. 
Do you have time to have a look at them?

martin-fleck-at commented:
@ivy-lli Interesting, yes I'll have a look at them. Is it simply enough to download the latest nightly Designer?

ivy-cst commented:
> @ivy-lli Interesting, yes I'll have a look at them. Is it simply enough to download the latest nightly Designer?

Yes

martin-fleck-at commented:
@ivy-cst @ivy-lli I fixed the two test cases and some deprecation warnings. However, I fixed the test cases by adapting the expected sizes so it would be great if you could have a look. I added comments in the test case to explain why I think the numbers are correct but maybe there is some other expected behavior that I'm missing.

ivy-lli commented:
Looks good, but it seems that now other tests are failing... Maybe because of these deprecation fixes? 🙈

ivy-lli commented:
But honestly, I'm also not sure why those are now broken, because locally they run for me. For you too?
Maybe it is problem with our build infra / setup / whatever.

martin-fleck-at commented:
@ivy-lli Interesting, I just let them run locally twice and they all go through without any problems. Can we maybe retrigger the CI?

ivy-lli commented:
@martin-fleck-at Ah I understand now the problem.
The viewer integration tests are now failing. They never run before, because the build was always interrupted by the failing standalone integration tests.

The problem is that currently the viewer is not working at all. It complains about that no ChangeBoundsManager is registered. However, we excluded it because we don't want that someone can move elements around in the viewer, but it seems that after the update it is now required. 
Maybe we should implement a fake ChangeBoundsManager which does nothing?

ivy-lli commented:
I was able to fix it with the following:
*viewer/src/tools/di.config.ts*
```ts
import { ChangeBoundsManager, FeatureModule, bindAsService, changeBoundsToolModule } from '@eclipse-glsp/client';
import { injectable } from 'inversify';

export const ivyViewerChangeBoundsToolModule = new FeatureModule(
  (bind, unbind, isBound, rebind) => {
    const context = { bind, unbind, isBound, rebind };
    // GLSP replacements
    bindAsService(context, ChangeBoundsManager, IvyChangeBoundsManager);
  },
  { featureId: changeBoundsToolModule.featureId }
);

@injectable()
export class IvyChangeBoundsManager extends ChangeBoundsManager {}
```

*viewer/src/di.config.ts*
```ts
const container = createIvyDiagramContainer(
  ...
  { replace: ivyViewerChangeBoundsToolModule },
  ...
);
```

martin-fleck-at commented:
@ivy-lli Ah, you are absolutely right, I can now also reproduce this locally! It seems that the `NodeCreationTool` needs the change bounds manager so that we can move the ghost element as feedback. I checked but I don't think node creation is possible in the viewer, is that right? At least when I simply remove the `nodeCreationToolModule`, the tests run through. I pushed another commit that should fix that. However, if node creation should be possible, we will need your fix.

ivy-lli commented:
No your right node creation is not needed, so remove it should definitely be fine!
I've triggered the builds.

ivy-lli commented:
It seems that they were successful. So we can merge right?

martin-fleck-at commented:
@ivy-lli Yes, I'll open another PR as soon as the GLSP release is out and I'll make the final adaptations then but I think this PR is large enough as it is ;-)

martin-fleck-at commented:
@ivy-lli Very interesting, I hadn't noticed, sorry for that! I saw that you already opened a PR to fix the decoration markers (https://github.com/axonivy/process-editor-client/pull/571). If that fixes that issue, would it be okay to double check the warnings once I upgrade to the release?

ivy-lli commented:
No problem :)
reading is hard. at least for the override warnings -> use the new `overrideModelElement` function. However I need to move the marker to the diagram config. It seems that replace container module are defined first...

martin-fleck-at commented:
@ivy-lli When we use the `replace` container configuration, the module takes the place of the original module in the least whereas the `add` (even if combined with `remove`) always placed the module at the end which caused some problems.
